src/HtmlProcessor/AbstractHtmlProcessor.php

```diff
@@ -302,6 +308,111 @@ private function addContentTypeMetaTag(string $html): string
+     *
+     * @return bool
+     */
+    private function hasContentTypeMetaTagInHead(string $html): bool
+    {
+        $contentTypeMetaTagMatchCount = \preg_match(
+            '%<meta(?=\\s)[^>]*\\shttp-equiv=(["\']?+)Content-Type\\g{-1}[\\s/>]%i',
+            $html,
+            $matches,
+            PREG_OFFSET_CAPTURE
+        );
+        if (\is_int($contentTypeMetaTagMatchCount) && $contentTypeMetaTagMatchCount > 0) {
+            /** @psalm-var array<int, array{0:string, 1:int}> $matches */
+            $matchPosition = $matches[0][1];
+            $htmlBefore = \substr($html, 0, $matchPosition);
+            try {
+                $hasContentTypeMetaTagInHead = !$this->hasEndOfHeadElement($htmlBefore);
+            } catch (\RuntimeException $exception) {
+                // If something unexpected occurs, assume the `Content-Type` that was found is valid.
+                \trigger_error($exception->getMessage());
+                $hasContentTypeMetaTagInHead = true;
+            }
+        } else {
+            $hasContentTypeMetaTagInHead = false;
+        }
+
+        return $hasContentTypeMetaTagInHead;
+    }
+
+    /**
+     * Tests whether the `<head>` element ends within the given HTML.  Due to tag omission rules, HTML parsers are
+     * expected to end the `<head>` element and start the `<body>` element upon encountering a start tag for any element
+     * which is permitted only within the `<body>`.
+     *
+     * @param string $html
+     *
+     * @return bool
+     *
+     * @throws \RuntimeException
+     */
+    private function hasEndOfHeadElement(string $html): bool
+    {
+        $headEndTagMatchCount
+            = \preg_match('%<(?!' . static::TAGNAME_ALLOWED_BEFORE_BODY_MATCHER . '[\\s/>])\\w|</head>%i', $html);
+        if (\is_int($headEndTagMatchCount) && $headEndTagMatchCount > 0) {
+            // An exception to the implicit end of the `<head>` is any content within a `<template>` element, as well in
+            // comments.  As an optimization, this is only checked for if a potential `<head>` end tag is found.
+            $htmlWithoutCommentsOrTemplates = $this->removeHtmlTemplateElements($this->removeHtmlComments($html));
+            if ($htmlWithoutCommentsOrTemplates === $html) {
+                $hasEndOfHeadElement = true;
+            } else {
+                $hasEndOfHeadElement = $this->hasEndOfHeadElement($htmlWithoutCommentsOrTemplates);
+            }
+        } else {
+            $hasEndOfHeadElement = false;
+        }
+
+        return $hasEndOfHeadElement;
+    }
+
+    /**
+     * Removes comments from the given HTML, including any unclosed, for which the remainder of the string is removed.
+     *
+     * @param string $html
+     *
+     * @return string
+     *
+     * @throws \RuntimeException
+     */
+    private function removeHtmlComments(string $html): string
+    {
+        $result = \preg_replace('/<!--[^-]*+(?:-(?!->)[^-]*+)*+(?:-->|$)/', '', $html);
```

Comment:
Can we move the regexp to a well-named local variable (or constant)?

Reply:
Moved to a constant, and similaraly for the pattern used by `removeHtmlTemplateElements`.